### ISSUE TITLE
ES2015's object literal shorthand properties breaks `key-spacing`

### ISSUE THREAD

**Issue by mdziekon** [eslint/eslint#3678]
It appears that passing shorthand property to object literal breaks `key-spacing` rule:
```javascript
const xyz = 123;
const test = {
    aaaa: 123,
    bbbb: 1123,
    xyz
};
```

with these errors
```javascript
error  Missing space before value for key "aaaa"  key-spacing
error  Missing space before value for key "bbbb"  key-spacing
```

ESLint version: v1.3.1
``.eslintrc`` file: http://pastebin.com/0Z76EfZm

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by mdziekon** [comment]
False alarm, it appears I was just using an old version of ``babel-eslint`` (I had to update to version 4.0.0).
Sorry for any trouble.

**Comment by mdziekon** [comment]
If I'm right about this problem, this test case probably should be included as a valid example of this rule:
```javascript
{
    code: [
        "var a = 'a';",
        "var b = 'b';",
        "",
        "var test = {",
        "    prop: 123,",
        "    a,",
        "    b",
        "};"
    ].join("\n"),
    ecmaFeatures: { objectLiteralShorthandProperties: true },
    options: [{ "align": "value" }]
}
```

**Comment by bathos** [comment]
I can confirm seeing this too with the same eslintrc settings for this rule. It stops if there's a newline between shorthand props and traditional props. I'm not confident this can squarely be called a bug because it's ambiguous how this should be treated, and the current behavior could be seen as consistent with the rule description ("Groups of properties separated by blank lines are considered distinct and can have different alignment than other groups.")

**Comment by mdziekon** [comment]
@bathos fair enough, but maybe it could be explicitly said in docs about these groups in context of spread operator? I thing it would be more straightforward for newcomers to eslint just like me.

**Comment by nzakas** [comment]
@btmills ?

**Comment by bathos** [comment]
If this is incorrect behavior, I'd point out that it also applies to method shorthand:

    propA: 'abc',
    propB: 'def',
    method() {}

Lines 1 and 2 are flagged as violating key-spacing.

**Comment by btmills** [comment]
Just confirmed this. The rule was written before ES6 support was added to ESLint, so it needs to be updated to support ES6 shorthand properties and methods. The rule shouldn't ever warn on shorthand properties or methods. Shorthand properties should continue alignment groups. Methods covering a single line should continue alignment groups, and methods spanning multiple lines should separate groups, just as with function expression values.

**Comment by pcostanz** [comment]
Confirming that this is an issue using destructuring, no amount of spaces will make this work.

lint error:
```
  119:7  warning  Missing space after key "color"           key-spacing
  120:7  warning  Missing space after key "display"         key-spacing
  121:7  warning  Missing space after key "fontFamily"      key-spacing
```

snippet:
```
{
      ...TYPE.HEADING_SMALL,
      color      : COLOR_NEUTRAL['200'],
      display    : 'block',
      fontFamily : 'Roboto, sans-serif'
}
```

**Comment by nzakas** [comment]
Working on this.

**Comment by nzakas** [comment]
No longer working on this. I couldn't figure out how this rule works. @btmills we really need some better comments and refactoring of this rule.

**Comment by btmills** [comment]
@nzakas yeah, it's gotten really complicated as parts have been added. Once answers to the currently open issues are nailed down, I'll probably rewrite large chunks of it and fix everything all at once.

**Comment by mysticatea** [comment]
I'm interested in separating the rule to `key-colon-spacing` and such as `column-alignment`.
But I don't have enough time now...

**Comment by btmills** [comment]
:+1: I second the motion; good idea. Not sure about the names though. What about property-colon-spacing and property-alignment?

**Comment by nzakas** [comment]
Is it possible to split these? I thought we looked before and determined it wasn't possible?

**Comment by mysticatea** [comment]
@nzakas I'm sorry for late. I thought it's possible: https://gist.github.com/mysticatea/27f32288f1ac024c88af

- This `property-colon-spacing` has an option `onlyLongestKey`.
- This `property-alignment` checks only positions which are between a property and the next property.

In this way, it can divide the work of `key-spacing`.
But this idea was thought before autofix arrived, so not autofix friendly.
I guess autofix of `key-spacing` arranges positions to the property of the longest key. So this separation idea may be bad.

**Comment by damonmcminn** [comment]
FWIW, as I liberally use object literal shorthand notation, the `key-spacing` rule in my `.eslintrc` has to be commented out as:

1. a group containing shorthand and regular notations will always fail
2. separating them into distinct groups *and* using `--fix` incorrectly "fixes" the code

**Comment by nzakas** [comment]
@damonmcminn can you file a separate issue about your trouble with `--fix`?

**Comment by damonmcminn** [comment]
@nzakas will do

**Comment by IanVS** [comment]
Where did we leave this?  Do we want to try to split them out, or no?

**Comment by nzakas** [comment]
Up to @btmills.

**Comment by mincedmit** [comment]
```
        return React.createElement(cls, {
          dispatch: this.context.store.dispatch,
          ...this.props,
          ...this.state
        });
``` as an example still breaks .. the fix doesn't seem to work with the spread operator, unless I'm misinterpreting the rule that should be valid

```
        return React.createElement(cls, {
          dispatch: this.context.store.dispatch,
          test,
          anotherTest
        });
``` is fine now

current workaround to pass linting is
```
return React.createElement(cls, {
          ...{
            dispatch: this.context.store.dispatch
          },
          ...this.props,
          ...this.state
        });
```

**Comment by btmills** [comment]
Looks like my fix only addressed ES6. @mincedmit can you file a new issue for that?

**Comment by bathos** [comment]
@btmills I think that issue might fall under the purview of babel-eslint / eslint-plugin-babel.

**Comment by btmills** [comment]
@bathos why's that? We have experimental support for object spread in the default Espree parser, so that syntax should be supported.

**Comment by bathos** [comment]
Oh -- for some reason I thought eslint only concerned itself with the official spec and left everything else to plugins, but I don't know how I came to believe that so it may well be untrue.

**Comment by btmills** [comment]
In every other case you'd be exactly right, but we made an exception for `experimentalObjectRestSpread` to enable usage with React. See eslint/espree#163 if you're interested.

**Comment by davidmaxwaterman** [comment]
I'm seeing this same error for reassignment during destructuring, as done here :

https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assigning_to_new_variable_names

I accept that it might be considered appropriate to add a space after the colon, as in the case of a key/value pair, but the wording of the error is a little misleading since it isn't a key/value pair. Is it worth rewording the error message, or at least adding such an example to the docs so people can see it is not a mistake?

**Comment by btmills** [comment]
@davidmaxwaterman can you file a new issue and tag me in it? The fields in the issue template will help me see what's going on.
